File: core/x/javaapi/api/src/main/java/viaduct/java/api/context/RootFieldCall.java
package viaduct.java.api.context;

import viaduct.apiannotations.ExperimentalApi;
import viaduct.apiannotations.InternalApi;
import viaduct.java.api.reflect.RootObjectField;
import viaduct.java.api.types.Arguments;
import viaduct.java.api.types.GraphQLObject;

/**
 * Describes a root field and the arguments used to call it.
 *
 * <p>Generated root field methods return this value without executing the field. A resolver passes
 * it to {@link ResolverExecutionContext#ref} to create a lazy reference that the engine resolves
 * later.
 *
 * @param <T> the field's object output type
 */
@ExperimentalApi
public interface RootFieldCall<T extends GraphQLObject> {

  @InternalApi
  RootObjectField<?, T, ? extends Arguments> field();

  /**
   * Builds the arguments the field is called with, using {@code context} to convert input values.
   */
  @InternalApi
  Arguments arguments(ExecutionContext context);
}

File: core/tenant/api/src/main/kotlin/viaduct/api/context/RootFieldCall.kt
package viaduct.api.context

import viaduct.api.reflect.RootObjectField
import viaduct.api.types.Arguments
import viaduct.api.types.Object
import viaduct.apiannotations.ExperimentalApi
import viaduct.apiannotations.InternalApi

/**
 * Describes a root field and the arguments used to call it.
 *
 * Generated root field functions return this value without executing the field. A resolver passes
 * it to [ResolverExecutionContext.ref] to create a lazy reference that the engine resolves later.
 */
@ExperimentalApi
interface RootFieldCall<T : Object> {
    @InternalApi
    fun field(): RootObjectField<*, T, Arguments>

    /** Builds the arguments the field is called with, using [context] to convert input values. */
    @InternalApi
    fun arguments(context: ExecutionContext): Arguments
}
